### ISSUE TITLE
[BUG] Test Coverage Varies Between Builds

### ISSUE THREAD

**Issue by PSchmiedmayer** [Apodini/Apodini#152]
**Describe the bug**
The test coverage of [`Sources/Apodini/Semantic Model Builder/Model/EndpointPath.swift`](https://github.com/Apodini/Apodini/blob/develop/Sources/Apodini/Semantic%20Model%20Builder/Model/EndpointPath.swift#L72) seems to fluctuate between builds. Sometimes line 72 (`return false`) of the `EndpointPath: Equatable` extension is not covered and executed when running the test suite.
```swift
extension EndpointPath: Equatable {
    static func == (lhs: EndpointPath, rhs: EndpointPath) -> Bool {
        switch (lhs, rhs) {
        case (.root, .root):
            return true
        case let (.string(lhsPath), .string(rhsPath)):
            return lhsPath == rhsPath
        case let (.parameter(lhsParameter), .parameter(rhsParameter)):
            return lhsParameter.id == rhsParameter.id
        default:
            return false
        }
    }
}
```
The default case seems to be executed sometimes while sometimes the test coverage and a breakpoint at the line seems to indicate that the line is not executed on some test runs.

**To Reproduce**

1. Set a breakpoint on the default case of the `EndpointPath: Equatable` extension.
2. Run the test suite multiple times.
3. Sometimes the line is executed, sometimes it is not executed.

**Expected behavior**
The test coverage shouldn't change between executions of the test suite and the behavior of the Apodini code should be deterministic.

**Environment**

 - Swift Version: 5.3
 - Xcode Version: 12.3 (12C33)

**Comment by lschlesinger** [comment]
The test with non-deterministic behavior seems to be [`SharedSemanticModelBuilderTests.testEndpointsTreeNodes`](https://github.com/Apodini/Apodini/blob/develop/Tests/ApodiniTests/SemanticModel/SharedSemanticModelBuilderTests.swift#L106).
I think the problem is in [this line](https://github.com/Apodini/Apodini/blob/develop/Sources/Apodini/Semantic%20Model%20Builder/Model/Endpoint.swift#L235).

My guess would be that this is due to the dictionary (which is inherently an unordered collection) that we use for [`nodeChildren: [EndpointPath: EndpointsTreeNode]` in `EndpointsTreeNode`](https://github.com/Apodini/Apodini/blob/develop/Sources/Apodini/Semantic%20Model%20Builder/Model/Endpoint.swift#L165).
Since it is unordered, the lookup with any [_next_ `EndpointPath`](https://github.com/Apodini/Apodini/blob/develop/Sources/Apodini/Semantic%20Model%20Builder/Model/Endpoint.swift#L235)s will non-deterministically compare to other available keys inside the dictionary; hence, depending on this non-deterministic order, the `default` case will be executed or not.

If we want to preserve the order of `nodeChildren` (at the cost of slower key lookup), we could probably use [`KeyValuePairs`](https://developer.apple.com/documentation/swift/keyvaluepairs) instead?

**Comment by PSchmiedmayer** [comment]
@lschlesinger Thanks for the fast response!! Your explanation sounds very plausible to me! A great catch, the unordered collection would be a good explanation for the phenomenon and using `KeyValuePairs` indeed seems like a good solution to make the lookup deterministic. Always great to see what great types are found in the Swift Standard Library 🚀

As summarized by you, the issue doesn't seem to be a non deterministic behavior in Apodini but more a non-deterministically comparison that is out of our hands. 
The alternative of using `KeyValuePairs` might be to just add a test case that covers all cases and guarantees a complete test coverage independent of the implementation of getting a next `EndpointsTreeNode`.
I personally would lean to keeping the implementation as it is and just adding an additional test case so we are not making design deceptions just for test coverage now that we know that this is not really "our fault".

But I would also be open about chaining up the implementation. What would your opinion about this @lschlesinger?

**Comment by lschlesinger** [comment]
> @lschlesinger Thanks for the fast response!! Your explanation sounds very plausible to me! A great catch, the unordered collection would be a good explanation for the phenomenon and using `KeyValuePairs` indeed seems like a good solution to make the lookup deterministic. Always great to see what great types are found in the Swift Standard Library 🚀
> 
> As summarized by you, the issue doesn't seem to be a non deterministic behavior in Apodini but more a non-deterministically comparison that is out of our hands.
> The alternative of using `KeyValuePairs` might be to just add a test case that covers all cases and guarantees a complete test coverage independent of the implementation of getting a next `EndpointsTreeNode`.
> I personally would lean to keeping the implementation as it is and just adding an additional test case so we are not making design deceptions just for test coverage now that we know that this is not really "our fault".
> 
> But I would also be open about chaining up the implementation. What would your opinion about this @lschlesinger?

I would also prefer adding tests cases and created a PR for that #157.

**Comment by lschlesinger** [comment]
fixed with #157
